File: voc-mesh-platform/packages/voc-schema/src/sensor.ts
import { z } from "zod";

export const SensorType = z.enum(["ENOSE", "GROUND", "ATMOSPHERE"]);
export type SensorType = z.infer<typeof SensorType>;

export const SensorSchema = z.object({
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  parcela_id: z.string().uuid(),
  hardware_id: z.string(),
  sensor_type: SensorType,
  lat: z.number(),
  lng: z.number(),
  active: z.boolean(),
  last_seen: z.string().optional(),
  battery_pct: z.number().optional(),
  signal_rssi: z.number().optional(),
});
export type Sensor = z.infer<typeof SensorSchema>;

export const VOCReadingSchema = z.object({
  time: z.string(),
  sensor_id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  compound: z.string(),
  value_ppm: z.number(),
  raw_adc: z.number(),
  temperature_c: z.number(),
  humidity_pct: z.number(),
});
export type VOCReading = z.infer<typeof VOCReadingSchema>;

export const SensorMessageSchema = z.object({
  hardware_id: z.string(),
  timestamp_ms: z.number(),
  temperature_c: z.number(),
  humidity_pct: z.number(),
  readings: z.array(
    z.object({
      compound: z.string(),
      value_ppm: z.number(),
      raw_adc: z.number(),
    })
  ),
  pm25: z.number(),
  pm10: z.number(),
});
export type SensorMessage = z.infer<typeof SensorMessageSchema>;

File: voc-mesh-platform/apps/api/app/models/sensor.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Sensor(Base):
    __tablename__ = "sensors"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False, index=True
    )
    parcela_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("parcelas.id"), nullable=True
    )
    hardware_id: Mapped[str] = mapped_column(String(128), unique=True, nullable=False)
    sensor_type: Mapped[str] = mapped_column(String(64), nullable=False)
    lat: Mapped[float | None] = mapped_column(Float, nullable=True)
    lng: Mapped[float | None] = mapped_column(Float, nullable=True)
    active: Mapped[bool] = mapped_column(Boolean, default=True)
    last_seen: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    firmware_version: Mapped[str | None] = mapped_column(String(32), nullable=True)


class VOCReading(Base):
    __tablename__ = "voc_readings"

    time: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        primary_key=True,
        default=lambda: datetime.now(timezone.utc),
    )
    sensor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sensors.id"), primary_key=True
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False, index=True
    )
    compound: Mapped[str] = mapped_column(String(64), nullable=False)
    value_ppm: Mapped[float] = mapped_column(Float, nullable=False)
    raw_adc: Mapped[int | None] = mapped_column(Integer, nullable=True)
    temperature_c: Mapped[float | None] = mapped_column(Float, nullable=True)
    humidity_pct: Mapped[float | None] = mapped_column(Float, nullable=True)
